File: BinarySearch/C++/153FindMinimuminRotatedSortedArray.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int left {0};
        int right = nums.size() - 1;
        int result = nums[0];


        // Start binary search 
        while (left <= right){

            // Condition if array is not rotated
            if (nums[left] < nums[right])
            {
                result = min(result, nums[left]);
                break;
            }

            int mid = ((right - left) / 2) + left;
           
            // Update the min value
            result = std::min(result, nums[mid]);

            // if midpoint val  is greater than right val, we serach target is in the right side of
            // the array to the midpoint i.e [mid+1, r] other wise we search in the left part of the array 
            // i.e [l, mid-1]
            if (nums[mid] > nums[right]){
                left = mid + 1;
            } else {
                right = mid - 1;
            }   
        }
        
        return result;
        
    }
};

File: BinarySearch/Python/153FindMinimuminRotatedSortedArray.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1
        res = nums[l]

        # Start binary seach
        while l <= r:
            # if array is aleady sorted, break the loop
            if nums[l] < nums[r]:
                return min(nums[l], res)
                break
                
            m = (l + r) // 2
            #If nums[m] is greater than nums[r], then the minimum value must
            #lie in the range [m+1, r]. So, res is updated with min(res, nums[m+1]).
            if nums[m] > nums[r]:
                l = m + 1
                res = min(res, nums[m + 1])
            #f nums[m] is not greater than nums[r], then the minimum value may lie in the range [l, m]. 
            #So, res is updated with min(res, nums[m]).
            else:
                r = m - 1
                res = min(res, nums[m])
        return res